Port cloudwatch exporter to Flow (#4169)
grafana/agent: docs/sources/flow/reference/components/prometheus.exporter.cloudwatch.md

---
# NOTE(thepalbi, from rfratto): the title below has zero-width spaces injected into it to
# prevent it from overflowing the sidebar on the rendered site. Be careful when
# modifying this section to retain the spaces.
#
# Ideally, in the future, we can fix the overflow issue with css rather than
# injecting special characters.

title: prometheus.exporter.cloudwatch
---

# prometheus.exporter.cloudwatch

The `prometheus.exporter.cloudwatch` component embeds

> 

Reviewer: extra blank line

```
"ec2:DescribeTags",
"ec2:DescribeInstances",
"ec2:DescribeRegions",
"ec2:DescribeTransitGateway*"
```

The following IAM permission is required to discover tagged [API Gateway](https://aws.amazon.com/es/api-gateway/) REST APIs:

```
"apigateway:GET"
```

The following IAM permissions are required to discover tagged [Database Migration Service](https://aws.amazon.com/dms/) (DMS) replication instances and tasks:

```
"dms:DescribeReplicationInstances",
"dms:DescribeReplicationTasks"
```

To use all of the integration features, use the following AWS IAM Policy:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Stmt1674249227793",
      "Action": [
        "tag:GetResources",
        "cloudwatch:GetMetricData",
        "cloudwatch:GetMetricStatistics",
        "cloudwatch:ListMetrics",
        "ec2:DescribeTags",
        "ec2:DescribeInstances",
        "ec2:DescribeRegions",
        "ec2:DescribeTransitGateway*",
        "apigateway:GET",
        "dms:DescribeReplicationInstances",
        "dms:DescribeReplicationTasks"
      ],
      "Effect": "Allow",
      "Resource": "*"
    }
  ]
}
```

## Usage

```river

> prometheus.exporter.cloudwatch "queues" {

Reviewer: Could you copy this snippet to a file and run `riverfmt` on it?

## metric block

Represents an AWS Metrics to scrape. To see available metrics, AWS does not keep a documentation page with all available metrics.
Follow [this guide](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/viewing_metrics_with_cloudwatch.html) on how to explore metrics, to easily pick the ones you need.

| Name         | Type           | Description                                                      | Default | Required |
| ------------ | -------------- | ---------------------------------------------------------------- | ------- | -------- |
| `name`       | `string`       | Metric name.                                                     |         | yes      |
| `statistics` | `list(string)` | List of statistics to scrape. Ex: `"Minimum"`, `"Maximum"`, etc. |         | yes      |
| `period`     | `duration`     | See [period][] section below.                                    |         | yes      |

[period]: #period

## role block

Represents an [AWS IAM Role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html). If omitted, the AWS role that corresponds to the credentials configured in the environment will be used.

> Multiple roles can be useful when scraping metrics from different AWS accounts with a single pair of credentials. In this case, a different role
> is configured for the agent to assume prior to calling AWS APIs, therefore, the credentials configured in the system need
> permission to assume the target role. See [this documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_permissions-to-switch.html) on how to configure this.

Reviewer 2: ```suggestion
Multiple roles can be useful when scraping metrics from different AWS accounts with a single pair of credentials. In this case, a different role
is configured for the agent to assume before calling AWS APIs. Therefore, the credentials configured in the system need
permission to assume the target role. See [Granting a user permissions to switch roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_permissions-to-switch.html) in the AWS IAM documentation for more information about how to configure this.
```

| Name          | Type     | Description                                                           | Default | Required |
| ------------- | -------- | --------------------------------------------------------------------- | ------- | -------- |
| `role_arn`    | `string` | AWS IAM Role ARN the exporter should assume to perform AWS API calls. |         | yes      |
| `external_id` | `string` | External ID used when calling STS AssumeRole API. See [details][].    | `""`    | no       |

[details]: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html

## period

> Period controls how far back in time CloudWatch metrics are considered, during each agent scrape. We can split how these
> settings affects the produced values in two different scenarios.

Reviewer 2: ```suggestion
Period controls how far back in time CloudWatch metrics are considered during each agent scrape. We can split how these
settings affect the produced values in two different scenarios.
```

If all metrics within a job (discovery or static) have the same `Period` value configured, CloudWatch APIs will be requested
for metrics from the scrape time, to `Periods` seconds in the past. The values of these are exported to Prometheus.

![](https://grafana.com/media/docs/agent/cloudwatch-single-period-time-model.png)

On the other hand, if metrics with different `Periods` are configured under an individual job, this works differently.
First, two variables are calculated aggregating all periods: `length`, taking the maximum value of all periods, and
the new `period` value, taking the minimum of all periods. Then, CloudWatch APIs will be requested for metrics from

> `now - length` to `now`, aggregating each in samples for `period` seconds. For each metrics, the most recent sample

Reviewer 2: ```suggestion
`now - length` to `now`, aggregating each in samples for `period` seconds. For each metric, the most recent sample
```

3. `metric`: Metric name and statistics.

For example, if you want to scrape the same metrics in the discovery example, but for a specific AWS EC2 instance:

```river
prometheus.exporter.cloudwatch "static_instances" {
	sts_region = "us-east-2"

	static "instances" {
		regions    = ["us-east-2"]
		namespace  = "AWS/EC2"
		dimensions = {
			"InstanceId" = "i01u29u12ue1u2c",
		}

		metric {
			name       = "CPUUsage"
			statistics = ["Sum", "Average"]
			period     = "1m"
		}
	}
}
```

You can configure the `static` block one or multiple times to scrape metrics with different sets of `dimensions`.

| Name          | Type           | Description                                                                                                                                                  | Default | Required |
| ------------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------- | -------- |

> | `name`        | `string`       | Identifier of the static scraping job. When exported to Prometheus format corresponds to the `name` label.                                                   |         | yes      |

Reviewer 3: I may have misunderstood, but from what I gathered the `name` for `static` blocks is [derived from the block's label](https://github.com/grafana/agent/blob/2787f0faa6dd74aaf3af9df26338077fa11f055e/component/prometheus/exporter/cloudwatch/config.go#L51) (so `static "my_static_block" { }` would have `name == "my_static_block"`). I'm not aware of any other component blocks that do this, and not personally opposed to it, but I think it may be a bit confusing to add it to the arguments of the `static` block as it isn't an argument, per se. Should the description be moved as a note underneath the arguments table? Something like:
```markdown
A `static` block's label corresponds to the `name` label when exported to the Prometheus format.
```

Maybe @clayton-cornell and/or @rfratto can chime in on thoughts for this?

Reviewer 4: Labels for blocks shouldn't be listed as an argument. For an example of what other components do, see the [profile.custom block in pyroscope.scrape](https://grafana.com/docs/agent/latest/flow/reference/components/pyroscope.scrape/#profilecustom-block).

The arguments table should be reserved for literal arguments which are provided by the user via the `ARGUMENT_NAME = ARGUMENT_VALUE` syntax, putting more things in the table (like blocks) might confuse the reader, as we're trying to make the table be indicative of how the user would configure that thing.

Author: Corrected following the pyroscope example.
